### DISCUSSION TITLE
ComputeThermalExpansionEigenstrain with temperature dependent coefficient

### DISCUSSION THREAD

**Discussion by eduardogrodriguez89** [idaholab/moose#26996] [Q&A Modules: Solid mechanics] [answered]
Hi
I am trying to compute thermal expansion eigenstrain $\epsilon = \alpha  (T - T_{stress} )$ to determine the displacement of a bar in 1D.
The bar is fixed on the left side and the temperature is increased to determine the displacement on the right side caused by temperature.

If a determine the displacement manually using:
$\delta L = \alpha  L(T - T_{stress} )$   and   $\alpha$ = 12e-6 $ºC^{-1}$,  L = 10cm ,  dt=25 ºC
$\delta L$ = 0.003 cm

But when I calculate it with MOOSE, I get a different result 
$\delta L_{moose}$ = 0.0057cm

Postprocessor Values:
+----------------+----------------+----------------+----------------+----------------+
|time ----------------| disp_x_bar1 | ---------p1---------|--- strain_xx ---        | temperature    |
+----------------+----------------+----------------+----------------+----------------+
|   5.000000e+00 |   5.571429e-03 |   0.000000e+00 |   5.571429e-04 |   2.500000e+01 |
+----------------+----------------+----------------+----------------+----------------+

I would like to know if my code is ok or why I get these different results?

Here is the code and mesh 
[bar1_1d.zip](https://github.com/idaholab/moose/files/14511006/bar1_1d.zip)
[barra_1D_temp.zip](https://github.com/idaholab/moose/files/14511007/barra_1D_temp.zip)

```
[Mesh]
    [base]
        type = FileMeshGenerator
        file = 'bar1_1d.msh'
    []
[]
[GlobalParams]
     displacements = 'disp_x'
[]
[AuxVariables]
   [temp]
   []
[]
[Functions]
  [temperature_load]
    type = ParsedFunction
    expression =5.0*t
  []
[]
[Modules]
  [TensorMechanics]
    [Master]
      [all]
        strain = SMALL
        incremental = true
        add_variables = true
        eigenstrain_names = eigenstrain
        generate_output = 'strain_xx'
      []
    []
  []
[]
[AuxKernels]
  [tempfuncaux]
    type = FunctionAux
    variable = temp
    function = temperature_load
  []
[]
[BCs]
  [point_x]
    type = DirichletBC
    variable = disp_x
    boundary = p1
    value = 0.0
  []
[]
[Materials]
  [elasticity_tensor_1]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 120e6
    poissons_ratio = 0.3
    block = 'bar_1'
  []  
  [small_stress]
    type = ComputeFiniteStrainElasticStress
  []
  [thermal_expansion_strain_1]
    type = ComputeThermalExpansionEigenstrain
    stress_free_temperature = 0
    thermal_expansion_coeff = 12e-6
    temperature = temp
    eigenstrain_name = eigenstrain
    block = 'bar_1'
  []
[]
[Executioner]
  type = Transient
  solve_type = 'newton'
  line_search = none
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  l_max_its = 4
  l_tol = 1e-14
  nl_max_its = 30
  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-10
  start_time = 0.0
  end_time = 5
  dt = 0.1
  dtmin = 0.1
[]
[Outputs]
  exodus = true
[]
[Postprocessors]
  [strain_xx]
    type = ElementAverageValue
    variable = strain_xx
    execute_on = 'initial timestep_end'
  []
  [disp_x_bar1]
    type = PointValue
    point = '10 0 0'
    variable = disp_x
    execute_on = 'initial timestep_end'
  []
  [p1]
    type = PointValue
    point = '0 0 0'
    variable = disp_x
    execute_on = 'initial timestep_end'
  []
  [temperature]
    type = AverageNodalVariableValue
    variable = temp
  []
[] 

```

**Comment by GiudGiud** [comment]
Hello

It s going to be tied to how the load is applied. 
Why did you select `        incremental = true`?

@bwspenc in case I'm wrong

Guillaume

**Reply by eduardogrodriguez89** [reply]
I try to replicate the code in #25572

If I change incremental = false

I get the following  error

*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'strain_increment', requested by 'small_stress' is not defined on block bar_1
Material property 'rotation_increment', requested by 'small_stress' is not defined on block bar_1
Material property 'strain_increment', requested by 'small_stress_face' is not defined on block bar_1
Material property 'rotation_increment', requested by 'small_stress_face' is not defined on block bar_1
Material property 'strain_increment', requested by 'small_stress_neighbor' is not defined on block bar_1
Material property 'rotation_increment', requested by 'small_stress_neighbor' is not defined on block bar_1

**Reply by eduardogrodriguez89** [reply]
Why did you `select         incremental = true`?
@GiudGiud 
I also based my code on that example.
[https://mooseframework.inl.gov/source/materials/ComputeThermalExpansionEigenstrain.html](url)
(modules/solid_mechanics/test/tests/thermal_expansion/constant_expansion_coeff.i)

**Reply by GiudGiud** [reply]
Yeah this makes sense.
Thoughts @bwspenc on why displacement is 2x the analytical ?

**Reply by GiudGiud** [reply]
@eduardogrodriguez89 is the mesh converged?
If you add uniform refine = 1 in the mesh block does the result change

**Reply by eduardogrodriguez89** [reply]
The mesh converged.
I added uniform refine = 1 to the mesh block and I got the same result.

**Reply by eduardogrodriguez89** [reply]
I have done more simulations and a question has arisen.
In the postprocessor block, I have asked to display the eigenstrain_xx and the strain_xx. Since the example is only thermal, shouldn't the eigenstrain_xx be equal to strain_xx? Because I visualize that both are different.

thermal_eigenstrain = $\alpha  (T-T_{stress})$ $~~~~~~ \alpha$ = 12e-6  1/ºC , $(T-T_{stress})=30ºC$
thermal_eigenstrain = 3.600000e-04

Any recommendations with the code? 

```

Postprocessor Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | disp_x_p2      | eigenstrain_xx | p1             | strain_xx      | strain_yy      | stress_xx      | stress_yy      | temperature    |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+

|   3.000000e+00 |   4.680000e-03 |   3.600000e-04 |   0.000000e+00 |   4.680000e-04 |   4.680000e-04 |   4.952610e-07 |  -1.593754e-08 |   3.000000e+01 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
```
```
eigenstrain_xx = 3.600000e-04
strain_xx = 4.680000e-04
```



```
[Mesh]
  [bar]
    type = GeneratedMeshGenerator
    nx = 500
    ny=5
    ymax=0.1
    xmax = 10 
    dim = 2
  []
[]
[GlobalParams]
  displacements = 'disp_x disp_y'
[]
[AuxVariables]
  [temp]
    initial_condition = 0.0
  []
  [eigenstrain_xx]
    order = CONSTANT
    family = MONOMIAL
  []
[]
[Functions]
  [temperature_load]
    type = ParsedFunction
    expression ='10.0*(t+0)'
  []
[]
[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [all]
        strain = SMALL
        add_variables = true
        eigenstrain_names = eigenstrain
        generate_output = 'strain_xx stress_xx strain_yy stress_yy'
      []
    []
  []
[]
[AuxKernels]
  [tempfuncaux]
    type = FunctionAux
    variable = temp
    function = temperature_load
  []
  [eigenstrain_xx]
    type = RankTwoAux
    rank_two_tensor = eigenstrain
    variable = eigenstrain_xx
    index_i = 0
    index_j = 0
    execute_on = 'initial timestep_end'
  []
[]
[BCs]
  [point1_x]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0
  []
[]
[Materials]
  [elasticity_tensor_1]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 200e9
    poissons_ratio = 0.3
  []  
  [small_stress]
    type = ComputeLinearElasticStress
  []
  [thermal_expansion_strain_1]
    type = ComputeThermalExpansionEigenstrain
    stress_free_temperature = 0
    thermal_expansion_coeff = 12e-6
    temperature = temp
    eigenstrain_name = eigenstrain
  []
[]
[Executioner]
  type = Transient
  solve_type = 'newton'
  #line_search = none
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  l_max_its = 4
  l_tol = 1e-12
  nl_max_its = 30
  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-10
  start_time = 0.0
  end_time = 3
  dt = 0.1
  dtmin = 0.1
[]
[Outputs]
  exodus = true
[]
[Postprocessors]
  [strain_xx]
    type = ElementAverageValue
    variable = strain_xx
    execute_on = 'initial timestep_end'
  []
  [strain_yy]
    type = ElementAverageValue
    variable = strain_yy
    execute_on = 'initial timestep_end'
  []
  [stress_xx]
    type = ElementAverageValue
    variable = stress_xx
    execute_on = 'initial timestep_end'
  []
  [stress_yy]
    type = ElementAverageValue
    variable = stress_yy
    execute_on = 'initial timestep_end'
  []
  [disp_x_p2]
    type = PointValue
    point = '10 0 0'
    variable = disp_x
    execute_on = 'initial timestep_end'
  []
  [p1]
    type = PointValue
    point = '0 0 0'
    variable = disp_x
    execute_on = 'initial timestep_end'
  []
  [temperature]
    type = AverageNodalVariableValue
    variable = temp
  []
  [eigenstrain_xx]
    type = ElementAverageValue
    variable = eigenstrain_xx
    execute_on = 'initial timestep_end'
  []
[]
```

**Reply by GiudGiud** [reply]
@dschwen knows about eigenstrains

**Reply by jiangwen84** [reply, marked as answer]
@eduardogrodriguez89  You forgot the Poisson's ratio. The thermal strain is volumetric and you need to consider the Poisson's effect.   You can set poisson ratio = 0.0 to verify.

**Reply by eduardogrodriguez89** [reply]
It works. Thanks
